### PR TITLE
Migrated some remaining floats in simulation to integers (and decimals)

### PR DESCRIPTION
**Pull request by reaperrr**
Partially adresses #10845.

### CODE REVIEW

File: OpenRA.Game/WVec.cs [hunk 1]
```diff
@@ -82,8 +82,8 @@ public static WVec LerpQuadratic(WVec a, WVec b, WAngle pitch, int mul, int div)
 				return ret;
 
 			// Add an additional quadratic variation to height
-			// Uses fp to avoid integer overflow
-			var offset = (int)((float)(b - a).Length * pitch.Tan() * mul * (div - mul) / (1024 * div * div));
+			// Uses long to avoid integer overflow
```
Comment (pchote) [inline OpenRA.Game/WVec.cs]: This isn't a hot path, so we could probably get away with using `decimal` here (there's another similar line somewhere using long that could also change) and remove the risk of overflow completely.

Comment (reaperrr) [inline OpenRA.Game/WVec.cs]: I think I saw WPos use long in a similar case, should that change to decimal as well?

Comment (pchote) [inline OpenRA.Game/WVec.cs]: Yeah, I think that is the line I was referring to.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 2]
```diff
@@ -644,6 +644,71 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
...
+				if (engineVersion < 20160323)
+				{
+					if (node.Key == "SelfHealing")
+					{
+						var healIfBelowNode = node.Value.Nodes.FirstOrDefault(x => x.Key == "HealIfBelow");
+						if (healIfBelowNode != null)
+						{
+							// The HealIfBelow value is now an int percentage, so multiply the float with 100.
+							var oldValue = FieldLoader.GetValue<float>("HealIfBelow", healIfBelowNode.Value.Value);
+							healIfBelowNode.Value.Value = (oldValue * 100).ToString();
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: You should convert this to an int to truncate any fractional part before turning it to a string.

Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Likewise for the other conversions.

File: mods/cnc/maps/gdi06/rules.yaml [hunk 3]
```diff
@@ -163,7 +163,7 @@ RMBO.easy:
 		HP: 300
 	SelfHealing:
 		Delay: 10
-		HealIfBelow: 50%
```
Comment (pchote) [inline mods/cnc/maps/gdi06/rules.yaml]: Is this the last use of the `%` syntax for loading floats?  We should drop that from FieldLoader if it is.

File: OpenRA.Mods.Common/Traits/SelfHealing.cs [hunk 4]
```diff
@@ -18,7 +18,8 @@ class SelfHealingInfo : UpgradableTraitInfo, Requires<HealthInfo>
 	{
 		public readonly int Step = 5;
 		public readonly int Delay = 5;
-		public readonly float HealIfBelow = .5f;
+		[Desc("Heal if current health is below this percentage of full health.")]
+		public readonly int HealIfBelow = 50;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SelfHealing.cs]: Do we want to rename these fields to include `Percent` in their name?  (I know we do this for some fields, but I'm not sure how widely that convention is used).

File: mods/cnc/maps/gdi06/rules.yaml (hunk 3 above)
Comment (reaperrr) [inline mods/cnc/maps/gdi06/rules.yaml]: AI still uses `%` for actor build decisions.

File: OpenRA.Mods.Common/Traits/SelfHealing.cs (hunk 4 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/SelfHealing.cs]: I'd say it's almost a tie between `Percent` and `Percentage` currently, plus a couple of cases like this with neither.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Game/WPos.cs [hunk 5]
```diff
@@ -65,7 +65,7 @@ public static WPos LerpQuadratic(WPos a, WPos b, WAngle pitch, int mul, int div)
 
 			// Add an additional quadratic variation to height
 			// Attempts to avoid integer overflow by keeping the intermediate variables reasonably sized
-			var offset = (int)(((((((long)(b - a).Length * mul) / div) * (div - mul)) / div) * pitch.Tan()) / 1024);
+			var offset = (int)(((((((decimal)(b - a).Length * mul) / div) * (div - mul)) / div) * pitch.Tan()) / 1024);
```
Comment (pchote) [inline OpenRA.Game/WPos.cs]: You can now drop most of the nested parenthesis, and the comment above will need to be updated.

Comment (pchote) [inline OpenRA.Game/WPos.cs]: Compare with the calculation in `WVec`.

File: OpenRA.Mods.Common/Traits/EmitInfantryOnSell.cs [hunk 6]
```diff
@@ -59,8 +59,14 @@ void Emit(Actor self)
 			var health = self.TraitOrDefault<Health>();
 			var dudesValue = info.ValuePercent * cost;
 			if (health != null)
-				dudesValue = dudesValue * health.HP / health.MaxHP;
-			dudesValue /= 100;
+			{
+				var healthPercentage = 100 * health.HP / health.MaxHP;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EmitInfantryOnSell.cs]: You can avoid rounding issues near the threshold by doing:
```csharp
if (100 * health.HP >= info.MinHpPercentage * health.MaxHP)
    dudesValue = health.HP * dudesValue / health.MaxHP;
else
    dudesValue = 0;
```

**Comment by pchote** [comment]
Looks good otherwise :+1:

**Comment by reaperrr** [comment]
Updated.

**Comment by Mailaender** [comment]
Needs a rebase.

**Comment by reaperrr** [comment]
Rebased.
Also added one last commit that moves all upgrade rules from both PRs to the same `engineVersion`.

**Comment by Mailaender** [comment]
Thanks!

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3042ea078f707eba41d50645cf4bf6790e931634...1859227b12a757093693056d5012051c1950e8eb)

**Pull request merged**